### DISCUSSION TITLE
How are you deploying Quire?

### DISCUSSION THREAD

**Discussion by Erin-Cecele** [thegetty/quire#94] [Development]
We have recently released a new section in our documentation, [*Deploy Your Project*](https://quire.getty.edu/documentation/site-deploy/). In this, we outline Netlify and GitHub Pages as great options for previewing and hosting Quire sites. We are also exploring Amazon's S3 as another option. We'd love to hear what other deployment methods are being used in the community. This will enable us to provide additional recommendations, guidance, and support to our users. Thank you! 

- Are you using existing internal/institutional servers or something external?
- If internal, what’s the process like for you?
- If external, what host/service are you using, and what’s the deployment process like with them?
- On a scale of 1-10 (with 10 being the hardest), how easy is it to use?
<br>

@1000camels, @kjell, @garrettdashnelson, @chrisdaaz, @halfempty, @swroberts, @tiffanysprague

**Comment by swroberts** [comment]
Hi there!

We at Pitts Theology Library are still reviewing our deployment process, but we've got it down to a pretty simple sequence:

- In short, we deploy to an external server using a single SSH command.

- We use [Reclaim Hosting](https://reclaimhosting.com/) for many of our projects, so for Quire publications we have a folder on a shared server; this provides us basic usage statistics through the cPanel tool Awstats
> Side note: Reclaim can cost as little as $45 per year for hosting and a domain for small projects

- We have SSH access to the server, so we use an `rsync` command ([https://en.wikipedia.org/wiki/Rsync](https://en.wikipedia.org/wiki/Rsync)) to push the contents of a local folder to the correct server folder

- We like `rsync` because it only pushes files that have changed; this works well because Quire only updates the affected files in the "site" folder when the Markdown or CSS files change. As a result, after the first upload, each subsequent sync between local and server is quick and efficient

- Without `rsync` this process was probably a 5 on the difficulty scale because it required manual upload of files, often the entire batch because it was too difficult to filter out only changed files; after switching to `rsync` the process is closer to a 1.

- We haven't reviewed how well this approach works if more than one person is pushing changes to the server, but in theory, it could work but may be less efficient because `rsync` may overwrite all files if run from a different local machine

**Comment by chrisdaaz** [comment]
Thanks for sharing the update to the documentation. I deploy to AWS S3 and CloudFront manually (for no good reason, just haven't scripted it yet). I use the [AWS Command Line Interface](https://aws.amazon.com/cli/) to `sync` the `site` directory to a dedicated S3 bucket, then create a cache invalidation on CloudFront. This is probably an 8 or 9 on a scale of difficulty because I'm working from notes or memory. 

I'm also using a different `class:` for the `contents.md` page based on output format (`list` for PDF, `grid` for web), which is partly why deployments are manual and time consuming (maybe there's a configuration I'm failing to use for this?). 

With other static site projects, I prefer to deploy with GitHub Actions. If there's interest in create a GitHub Actions workflow for Quire, I'd be interested in helping with that.

**Reply by geealbers** [reply]
@chrisdaaz thanks for sharing this. For your different contents pages, you could create a page for each in the project and then use `online: false` and `pdf: false` in the page yaml to hide them from particular outputs. There's some info in the docs at  https://quire.getty.edu/documentation/pages/#hideshow-pages, but essentially you'd have one page that's like this for the PDF that wouldn't show up online:

```
type: contents
class: list
online: false
```
And one like this for online that wouldn't show up in the PDF:

```
type: contents
class: grid
pdf: false
```

**Comment by garrettdashnelson** [comment]
Hey all, sorry to be just noticing this thread now!

For our _Bending Lines_ Quire exhibition, I was in the deployment practice of building the site on my own machine, and then pushing it to a web server (locally hosted at the BPL) using `rsync`. 

With that said, since last spring I have been using Netlify for more and more projects, and it is terrific. I definitely think that if I do another Quire project, I'll probably set up CI to build it on Netlify, at least as a dev site if not for the final production.

One thing I will note still trips me up (not only about Quire about various other Hugo-built sites I've worked on) is handling relative paths if the final deploy will not be in the root directory of a domain or subdomain.

**Reply by geealbers** [reply]
Thanks @garrettdashnelson. I'm with you on the relative paths thing! We need to do more in Quire to figure that out and document it. One thing I've noticed with Hugo is that it does some very opinionated, and not always expected, things with URLs. I often get tripped up when trying to figure it out, but one thing in particular is that Hugo doesn't recognize or process URLs with its `relURL`/`absURL` functions or `relativeURLs`/`canonifyURLs` configs equally in all cases. For example, when I checked last, it was recognizing URLs inside `href` attributes but not when they're included in `style` attributes for say a background image.
